File: src/matlab/dft_half.m
function [y] = dft_half(x)
% Computation of the N/2-point DFT of a N-point real sequence (Recovery)
% Input:
% 	x - complex N/2-point sequence: x = DFT[ s(2k-1) + i*s(2k) ], where s is a N-point real-valued sequence
% Output:
%	y - complex recovered N-point sequence: y = DFT[s]
%
N=numel(x)*2;
w0=zeros(1, N/2);
w1=zeros(1, N/2);
y=zeros(1, N);

for k=1:N/2
	w0(k)=(1-1i*exp(-1i*2*pi*(k-1)/N))/2;
	w1(k)=(1+1i*exp(-1i*2*pi*(k-1)/N))/2;
end

x(N/2+1)=x(1);

for k=1:N/2
	y(k)= x(k)*w0(k)+conj(x(N/2-k+2))*w1(k);
end

y(N/2+1)=real(x(1))-imag(x(1));
for i=2:N/2
	y(N-i+2)=conj(y(i));
end

end

File: src/dft_half.c
/**
 * @file dft_half.c
 * @brief Computation of the N/2-point DFT of a N-point real sequence (Recovery)
 * @author matyunin.d
 * @date 28.08.2019
 */

#include <complex.h>
#include <math.h>

/**
 * @brief Recovery N-point sequence from N/2-point sequence
 * x(k) = N/2-point DFT[ s(2*k) + i*s(2*k+1) ] => y(k) = N-point DFT[ s(k) ]
 * s - N-point real-valued sequence
 * @param x Pointer to input array of N/2-point complex-valued sequence (length = N/2)
 * @param y Pointer to output recovered N-point complex-valued sequence (length = N)
 * @param N Length
 */
void dft_half(const double complex *x, double complex *y, unsigned int N)
{
	double complex w0, w1;

	w0 = (1 - I) / 2;
	w1 = (1 + I) / 2;
	y[0]= x[0] * w0 + conj(x[0]) * w1;

	for (unsigned int k = 1; k < N/2; k++) {
		w0 = (1 - I * cexp(-I * 2 * M_PI * k / N)) / 2;
		w1 = (1 + I * cexp(-I * 2 * M_PI * k / N)) / 2;
		y[k]= x[k] * w0 + conj(x[N/2-k]) * w1;
	}

	y[N/2] = creal(x[0]) - cimag(x[0]);
	for (unsigned int k = 1; k < N/2; k++)
		y[N-k] = conj(y[k]);
}


